File: swift/Sources/Rezel/Grammars/CppGrammar.swift
import Foundation

private let R = 82
private let L = 76
private let u = 117
private let U = 85
private let a = 97
private let z = 122
private let A = 65
private let Z = 90
private let Underscore = 95
private let Zero = 48
private let Quote = 34
private let ParenL = 40
private let ParenR = 41
private let Space = 32
private let Newline = 10
private let GreaterThan = 62

private func makeCppExternalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
	if name == "rawString" {
		let RawString = terms["RawString"]!
		return ExternalTokenizer { input, _ in
			// Raw string literals can start with: R, LR, uR, UR, u8R
			if input.next == L || input.next == U {
				input.advance()
			} else if input.next == u {
				input.advance()
				if input.next == Zero + 8 { input.advance() }
			}
			if input.next != R { return }
			input.advance()
			if input.next != Quote { return }
			input.advance()

			var marker = ""
			while input.next != ParenL {
				if input.next == Space || input.next <= 13 || input.next == ParenR { return }
				marker += String(UnicodeScalar(input.next)!)
				input.advance()
			}
			input.advance()

			let markerCodes = marker.unicodeScalars.map { Int($0.value) }
			while true {
				if input.next < 0 {
					input.acceptToken(RawString)
					return
				}
				if input.next == ParenR {
					var match = true
					var i = 0
					while match && i < markerCodes.count {
						if input.peek(i + 1) != markerCodes[i] { match = false }
						i += 1
					}
					if match && input.peek(markerCodes.count + 1) == Quote {
						input.acceptToken(RawString, endOffset: 2 + markerCodes.count)
						return
					}
				}
				input.advance()
			}
		}
	}

	if name == "fallback" {
		let templateArgsEndFallback = terms["templateArgsEndFallback"]!
		let MacroName = terms["MacroName"]!
		return ExternalTokenizer({ input, _ in
			if input.next == GreaterThan {
				// Provide a template-args-closing token when the next characters
				// are ">>", in which case the regular tokenizer will only see a
				// bit shift op.
				if input.peek(1) == GreaterThan {
					input.acceptToken(templateArgsEndFallback, endOffset: 1)
				}
			} else {
				// Notice all-uppercase identifiers
				var sawLetter = false
				var i = 0
				while true {
					if input.next >= A, input.next <= Z {
						sawLetter = true
					} else if input.next >= a, input.next <= z {
						return
					} else if input.next != Underscore, !(input.next >= Zero && input.next <= Zero + 9) {
						break
					}
					input.advance()
					i += 1
				}
				if sawLetter, i > 1 {
					input.acceptToken(MacroName)
				}
			}
		}, extend: true)
	}

	fatalError("Unknown CPP external tokenizer: \(name)")
}

private nonisolated(unsafe) let definition = hlTags["definition"] as! (Tag) -> Tag
private nonisolated(unsafe) let functionMod = hlTags["function"] as! (Tag) -> Tag
private nonisolated(unsafe) let specialMod = hlTags["special"] as! (Tag) -> Tag
private nonisolated(unsafe) let standardMod = hlTags["standard"] as! (Tag) -> Tag

nonisolated(unsafe) let cppHighlighting = styleTags([
	"typedef struct union enum class typename decltype auto template operator friend noexcept namespace using requires concept import export module __attribute__ __declspec __based": hlTags["definitionKeyword"] as Any,
	"extern MsCallModifier MsPointerModifier extern static register thread_local inline const volatile restrict _Atomic mutable constexpr constinit consteval virtual explicit VirtualSpecifier Access": hlTags["modifier"] as Any,
	"if else switch for while do case default return break continue goto throw try catch": hlTags["controlKeyword"] as Any,
	"co_return co_yield co_await": hlTags["controlKeyword"] as Any,
	"new sizeof delete static_assert": hlTags["operatorKeyword"] as Any,
	"NULL nullptr": hlTags["null"] as Any,
	"this": hlTags["self"] as Any,
	"True False": hlTags["bool"] as Any,
	"TypeSize PrimitiveType": standardMod(hlTypeName),
	"TypeIdentifier": hlTypeName,
	"FieldIdentifier": hlPropertyName,
	"CallExpression/FieldExpression/FieldIdentifier": functionMod(hlPropertyName),
	"ModuleName/Identifier": hlTags["namespace"] as Any,
	"PartitionName": hlTags["labelName"] as Any,
	"StatementIdentifier": hlTags["labelName"] as Any,
	"Identifier DestructorName": hlTags["variableName"] as Any,
	"CallExpression/Identifier": functionMod(hlTags["variableName"] as! Tag),
	"CallExpression/ScopedIdentifier/Identifier": functionMod(hlTags["variableName"] as! Tag),
	"FunctionDeclarator/Identifier FunctionDeclarator/DestructorName": functionMod(definition(hlTags["variableName"] as! Tag)),
	"NamespaceIdentifier": hlTags["namespace"] as Any,
	"OperatorName": hlTags["operator"] as Any,
	"ArithOp": hlTags["arithmeticOperator"] as Any,
	"LogicOp": hlTags["logicOperator"] as Any,
	"BitOp": hlTags["bitwiseOperator"] as Any,
	"CompareOp": hlTags["compareOperator"] as Any,
	"AssignOp": hlTags["definitionOperator"] as Any,
	"UpdateOp": hlTags["updateOperator"] as Any,
	"LineComment": hlTags["lineComment"] as Any,
	"BlockComment": hlTags["blockComment"] as Any,
	"Number": hlNumber,
	"String": hlString,
	"RawString SystemLibString": specialMod(hlString),
	"CharLiteral": hlTags["character"] as Any,
	"EscapeSequence": hlTags["escape"] as Any,
	"UserDefinedLiteral/Identifier": hlTags["literal"] as Any,
	"PreprocArg": hlTags["meta"] as Any,
	"PreprocDirectiveName #include #ifdef #ifndef #if #define #else #endif #elif": hlTags["processingInstruction"] as Any,
	"MacroName": specialMod(hlName),
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
	"< >": hlTags["angleBracket"] as Any,
	". ->": hlTags["derefOperator"] as Any,
	", ;": hlTags["separator"] as Any,
])

public nonisolated(unsafe) let cppParser: LRParser = {
	let externals: [String: TokenizerProtocol] = [
		"rawString": makeCppExternalTokenizer(name: "rawString", terms: CppParserData.termTable),
		"fallback": makeCppExternalTokenizer(name: "fallback", terms: CppParserData.termTable),
	]
	let spec = CppParserData.makeSpec(
		externals: externals,
		propSources: [cppHighlighting]
	)
	return LRParser(spec: spec)
}()

File: dotnet/Rezel/Grammars/CppGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class CppGrammar
{
    private const int R = 82;
    private const int L = 76;
    private const int U = 85;
    private const int u = 117;
    private const int A = 65;
    private const int Z = 90;
    private const int a = 97;
    private const int z = 122;
    private const int Underscore = 95;
    private const int Zero = 48;
    private const int Quote = 34;
    private const int ParenL = 40;
    private const int ParenR = 41;
    private const int Space = 32;
    private const int GreaterThan = 62;

    private static ITokenizer MakeCppExternalTokenizer(string name, Dictionary<string, int> terms)
    {
        if (name == "rawString")
        {
            var RawString = terms["RawString"];
            return new ExternalTokenizer((input, _) =>
            {
                if (input.Next == L || input.Next == U)
                {
                    input.Advance();
                }
                else if (input.Next == u)
                {
                    input.Advance();
                    if (input.Next == Zero + 8) input.Advance();
                }
                if (input.Next != R) return;
                input.Advance();
                if (input.Next != Quote) return;
                input.Advance();

                var marker = "";
                while (input.Next != ParenL)
                {
                    if (input.Next == Space || input.Next <= 13 || input.Next == ParenR) return;
                    marker += (char)input.Next;
                    input.Advance();
                }
                input.Advance();

                while (true)
                {
                    if (input.Next < 0)
                    {
                        input.AcceptToken(RawString);
                        return;
                    }
                    if (input.Next == ParenR)
                    {
                        var match = true;
                        for (var i = 0; match && i < marker.Length; i++)
                            if (input.Peek(i + 1) != marker[i]) match = false;
                        if (match && input.Peek(marker.Length + 1) == Quote)
                        {
                            input.AcceptToken(RawString, endOffset: 2 + marker.Length);
                            return;
                        }
                    }
                    input.Advance();
                }
            });
        }

        if (name == "fallback")
        {
            var templateArgsEndFallback = terms["templateArgsEndFallback"];
            var MacroName = terms["MacroName"];
            return new ExternalTokenizer((input, _) =>
            {
                if (input.Next == GreaterThan)
                {
                    if (input.Peek(1) == GreaterThan)
                        input.AcceptToken(templateArgsEndFallback, endOffset: 1);
                }
                else
                {
                    var sawLetter = false;
                    var i = 0;
                    for (; ; i++)
                    {
                        if (input.Next >= A && input.Next <= Z) sawLetter = true;
                        else if (input.Next >= a && input.Next <= z) return;
                        else if (input.Next != Underscore && !(input.Next >= Zero && input.Next <= Zero + 9)) break;
                        input.Advance();
                    }
                    if (sawLetter && i > 1) input.AcceptToken(MacroName);
                }
            }, extend: true);
        }

        throw new ArgumentException($"Unknown C++ external tokenizer: {name}");
    }

    private static readonly NodePropSource CppHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["typedef struct union enum class typename decltype auto template operator friend noexcept namespace using requires concept import export module __attribute__ __declspec __based"] = Tags.DefinitionKeyword,
        ["extern MsCallModifier MsPointerModifier extern static register thread_local inline const volatile restrict _Atomic mutable constexpr constinit consteval virtual explicit VirtualSpecifier Access"] = Tags.ModifierTag,
        ["if else switch for while do case default return break continue goto throw try catch"] = Tags.ControlKeyword,
        ["co_return co_yield co_await"] = Tags.ControlKeyword,
        ["new sizeof delete static_assert"] = Tags.OperatorKeyword,
        ["NULL nullptr"] = Tags.Null,
        ["this"] = Tags.Self,
        ["True False"] = Tags.Bool,
        ["TypeSize PrimitiveType"] = Tags.Standard(Tags.TypeName),
        ["TypeIdentifier"] = Tags.TypeName,
        ["FieldIdentifier"] = Tags.PropertyName,
        ["CallExpression/FieldExpression/FieldIdentifier"] = Tags.Function(Tags.PropertyName),
        ["ModuleName/Identifier"] = Tags.Namespace,
        ["PartitionName"] = Tags.LabelName,
        ["StatementIdentifier"] = Tags.LabelName,
        ["Identifier DestructorName"] = Tags.VariableName,
        ["CallExpression/Identifier"] = Tags.Function(Tags.VariableName),
        ["CallExpression/ScopedIdentifier/Identifier"] = Tags.Function(Tags.VariableName),
        ["FunctionDeclarator/Identifier FunctionDeclarator/DestructorName"] = Tags.Function(Tags.Definition(Tags.VariableName)),
        ["NamespaceIdentifier"] = Tags.Namespace,
        ["OperatorName"] = Tags.Operator,
        ["ArithOp"] = Tags.ArithmeticOperator,
        ["LogicOp"] = Tags.LogicOperator,
        ["BitOp"] = Tags.BitwiseOperator,
        ["CompareOp"] = Tags.CompareOperator,
        ["AssignOp"] = Tags.DefinitionOperator,
        ["UpdateOp"] = Tags.UpdateOperator,
        ["LineComment"] = Tags.LineComment,
        ["BlockComment"] = Tags.BlockComment,
        ["Number"] = Tags.Number,
        ["String"] = Tags.String,
        ["RawString SystemLibString"] = Tags.Special(Tags.String),
        ["CharLiteral"] = Tags.Character,
        ["EscapeSequence"] = Tags.Escape,
        ["UserDefinedLiteral/Identifier"] = Tags.Literal,
        ["PreprocArg"] = Tags.Meta,
        ["PreprocDirectiveName #include #ifdef #ifndef #if #define #else #endif #elif"] = Tags.ProcessingInstruction,
        ["MacroName"] = Tags.Special(Tags.Name),
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
        ["< >"] = Tags.AngleBracket,
        [". ->"] = Tags.DerefOperator,
        [", ;"] = Tags.Separator,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var termTable = CppParserData.TermTable!;
        var externals = new Dictionary<string, ITokenizer>
        {
            ["rawString"] = MakeCppExternalTokenizer("rawString", termTable),
            ["fallback"] = MakeCppExternalTokenizer("fallback", termTable),
        };
        var spec = CppParserData.MakeSpec(
            externals: externals,
            propSources: [CppHighlighting]
        );
        return new LRParser(spec);
    }
}
